### DISCUSSION TITLE
Inconsistent behavior in tests after TCA update with separated sharing

### DISCUSSION THREAD

**Discussion by mobilne** [pointfreeco/swift-sharing#42] [Q&A]
Hello,

after update on 1.17.0 and sharing lib 1.0.4 there are few changes I was forced to, like: 
```
extension PersistenceReaderKey where Self == PersistenceKeyDefault<InMemoryKey<User?>> {
    static var user: Self {
        PersistenceKeyDefault(.inMemory("user"), nil)
    }
}
```
to:
```
extension SharedReaderKey where Self == Sharing.InMemoryKey<User?>.Default {
    static var user: Self {
        Self[.inMemory("user"), default: nil]
    }
}
```

and in the reducer `state.user = newUser` to `state.$user.withLock { $0 = newUser }`

**Tests:**
Besides, some tests don't pass now, because direct assigning `@Shared(.user) var user = User(.Mocks.disabled)` does not put the desired value there, so I have to use withLock closure. But just in some tests.

Also when subscribe on shared in reducer, like:
```
          case .initializeCallLog:
              state.entries = mvProvider.retrieveRemovedContacts()
              return .publisher {
                  state.$user.publisher
                      .map(to: Action.userChanged)
              }
              .cancellable(id: CancelID.observeUser, cancelInFlight: true)

            case .userChanged:
                state.entries = mvProvider.retrieveRemovedContacts()
                return .none
```
the test complains that I don't handle userChanged action that is caused by emitting value on subscription. Before the test did not take it into account.

Example:
```
    func testInitializeCallLog() async {
        let mvService = User.MVService.Mocks.oneContact
        @Shared(.user) var user = User(mvService: mvService)

        let store = TestStore(
            initialState: CallLogCore.State(),
            reducer: CallLogCore.init
        )

        let retrieveExpectation = expectation(description: "retrieveRemovedContacts")
        store.dependencies.mvProvider.retrieveRemovedContacts = {
            retrieveExpectation.fulfill()
            return Contact.Mocks.oneContact
        }


        await store.send(.initializeCallLog)
        await fulfillment(of: [retrieveExpectation], timeout: 0.1)
        await store.skipInFlightEffects()
    }
```
This test now complaints on: API violation - multiple calls made to -[XCTestExpectation fulfill] for retrieveRemovedContacts. Because both .initializeCallLog and .userChanged actions use the dependency and .userChanged is called on shared user subscription.

Would you have any hint what I did wrong, in the past by using shared or now after update? The general question is why there is an emission of the action on shared state subscription and why the test did not reflect this before the update to 1.17.0, when sharing was part of the library...

Thanks. 
Sincerely,
David.

**Comment by mbrandonw** [comment]
Hi @mobilne, I believe all of these changes were covered in the migration guide. First, the [TCA migration guide for 1.17](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/migratingto1.17) mentions that if you want to maintain backwards compatibility with the version of sharing that existed in TCA prior to 1.17, you can pin to Swift Sharing 0.1.0. And then the [Swift Sharing migration guide for 1.0](https://swiftpackageindex.com/pointfreeco/swift-sharing/main/documentation/sharing/migratingto1.0) shows you how to upgrade to 1.x when you are ready. 

> after update on 1.17.0 and sharing lib 1.0.4 there are few changes I was forced to, like:

Yes, this is a 1.0 change. If you do not want to change your keys to this you can pin explicitly on 0.1 in order to maintain backwards compatibility.

> Besides, some tests don't pass now, because direct assigning @Shared(.user) var user = User(.Mocks.disabled) does not put the desired value there, so I have to use withLock closure. But just in some tests.

I am not able to reproduce this. Would you be able to create a simple, standalone test case that shows the behavior you are seeing?

> the test complains that I don't handle userChanged action that is caused by emitting value on subscription. Before the test did not take it into account.

This is also covered in the 1.0 migration guide [here](https://swiftpackageindex.com/pointfreeco/swift-sharing/main/documentation/sharing/migratingto1.0#Shared-publishers-now-immediately-emit-the-current-value). Previously the publisher acted more like a passthrough subject, now it behaves like a current value subject. Again, you can pin to 0.1 if you don't want this behavior for the time being.

**Reply by mobilne** [reply]
Hi, I did not notice this in the migration guide, thanks. Regarding the direct assigning the shared variables in tests, there is a fix in sharing library 1.1.1 Now It works.
